File: scripts/neo4j_memory_builder.js
/**
 * Neo4j Memory Builder
 * 
 * This script builds a knowledge graph of the projects directory using Neo4j memory commands.
 * It creates entities for projects, directories, and important files, and establishes relationships between them.
 */

const fs = require('fs');
const path = require('path');
const util = require('util');
const exec = util.promisify(require('child_process').exec);

// Configuration
const PROJECTS_DIR = '/home/thein/projects';
const IGNORE_DIRS = [
  'node_modules', '.git', '.vscode', 'out', 'dist', 'build', 'coverage', '.cache'
];
const IMPORTANT_FILES = [
  'package.json', 'README.md', '.augment-guidelines', 'tsconfig.json', 
  'docker-compose.yml', 'Dockerfile', 'devcontainer.json'
];
const MAX_BATCH_SIZE = 10; // Maximum number of entities/relations to create in a single batch

// Helper function to read file content safely
function readFileSafe(filePath, maxSize = 10000) {
  try {
    const stats = fs.statSync(filePath);
    if (stats.size > maxSize) {
      return `File too large (${stats.size} bytes)`;
    }
    return fs.readFileSync(filePath, 'utf8');
  } catch (error) {
    return `Error reading file: ${error.message}`;
  }
}

// Helper function to create entities in batches
async function createEntitiesBatch(entities) {
  if (entities.length === 0) return;
  
  for (let i = 0; i < entities.length; i += MAX_BATCH_SIZE) {
    const batch = entities.slice(i, i + MAX_BATCH_SIZE);
    try {
      await create_entities_memory({ entities: batch });
      console.log(`Created ${batch.length} entities (batch ${Math.floor(i / MAX_BATCH_SIZE) + 1})`);
    } catch (error) {
      console.error('Error creating entities:', error);
    }
  }
}

// Helper function to create relations in batches
async function createRelationsBatch(relations) {
  if (relations.length === 0) return;
  
  for (let i = 0; i < relations.length; i += MAX_BATCH_SIZE) {
    const batch = relations.slice(i, i + MAX_BATCH_SIZE);
    try {
      await create_relations_memory({ relations: batch });
      console.log(`Created ${batch.length} relations (batch ${Math.floor(i / MAX_BATCH_SIZE) + 1})`);
    } catch (error) {
      console.error('Error creating relations:', error);
    }
  }
}

// Process a project and add it to the knowledge graph
async function processProject(projectPath) {
  const projectName = path.basename(projectPath);
  console.log(`Processing project: ${projectName}`);
  
  const entities = [];
  const relations = [];
  
  // Create project entity
  const projectEntity = {
    name: projectName,
    entityType: 'Project',
    observations: [`Project directory: ${projectPath}`]
  };
  
  // Add package.json info if available
  const packageJsonPath = path.join(projectPath, 'package.json');
  if (fs.existsSync(packageJsonPath)) {
    try {
      const packageJson = JSON.parse(readFileSafe(packageJsonPath));
      if (packageJson.name) {
        projectEntity.observations.push(`Name: ${packageJson.name}`);
      }
      if (packageJson.description) {
        projectEntity.observations.push(`Description: ${packageJson.description}`);
      }
      if (packageJson.version) {
        projectEntity.observations.push(`Version: ${packageJson.version}`);
      }
    } catch (error) {
      console.error(`Error parsing package.json for ${projectName}:`, error.message);
    }
  }
  
  // Add README.md content if available
  const readmePath = path.join(projectPath, 'README.md');
  if (fs.existsSync(readmePath)) {
    const readmeContent = readFileSafe(readmePath);
    projectEntity.observations.push(`README: ${readmeContent}`);
  }
  
  entities.push(projectEntity);
  
  // Process important files
  for (const fileName of IMPORTANT_FILES) {
    const filePath = path.join(projectPath, fileName);
    if (fs.existsSync(filePath)) {
      const fileContent = readFileSafe(filePath);
      entities.push({
        name: `${projectName}:${fileName}`,
        entityType: 'File',
        observations: [`Path: ${filePath}`, `Content: ${fileContent}`]
      });
      
      relations.push({
        from: projectName,
        to: `${projectName}:${fileName}`,
        relationType: 'contains'
      });
    }
  }
  
  // Process .devcontainer directory if it exists
  const devcontainerDir = path.join(projectPath, '.devcontainer');
  if (fs.existsSync(devcontainerDir) && fs.statSync(devcontainerDir).isDirectory()) {
    entities.push({
      name: `${projectName}:devcontainer`,
      entityType: 'Component',
      observations: ['DevContainer configuration for the project']
    });
    
    relations.push({
      from: projectName,
      to: `${projectName}:devcontainer`,
      relationType: 'has'
    });
    
    // Process important devcontainer files
    const devcontainerFiles = ['devcontainer.json', 'docker-compose.yml', 'Dockerfile'];
    for (const fileName of devcontainerFiles) {
      const filePath = path.join(devcontainerDir, fileName);
      if (fs.existsSync(filePath)) {
        const fileContent = readFileSafe(filePath);
        entities.push({
          name: `${projectName}:devcontainer:${fileName}`,
          entityType: 'File',
          observations: [`Path: ${filePath}`, `Content: ${fileContent}`]
        });
        
        relations.push({
          from: `${projectName}:devcontainer`,
          to: `${projectName}:devcontainer:${fileName}`,
          relationType: 'contains'
        });
      }
    }
  }
  
  // Create entities and relations
  await createEntitiesBatch(entities);
  await createRelationsBatch(relations);
}

// Main function to build the knowledge graph
async function buildKnowledgeGraph() {
  console.log('Building knowledge graph of projects...');
  
  try {
    // Get list of projects
    const projects = fs.readdirSync(PROJECTS_DIR)
      .filter(item => {
        const itemPath = path.join(PROJECTS_DIR, item);
        return fs.statSync(itemPath).isDirectory() && !IGNORE_DIRS.includes(item);
      })
      .map(item => path.join(PROJECTS_DIR, item));
    
    console.log(`Found ${projects.length} projects: ${projects.map(p => path.basename(p)).join(', ')}`);
    
    // Process each project
    for (const project of projects) {
      await processProject(project);
    }
    
    console.log('Knowledge graph built successfully!');
  } catch (error) {
    console.error('Error building knowledge graph:', error);
  }
}

// Run the main function
buildKnowledgeGraph();

File: scripts/neo4j_memory_builder.py
#!/usr/bin/env python3
"""
Neo4j Memory Builder

This script builds a knowledge graph of the projects directory using Neo4j memory commands.
It creates entities for projects, directories, and important files, and establishes relationships between them.
"""

import os
import json
import sys
import time
from typing import List, Dict, Any, Optional

# Configuration
PROJECTS_DIR = '/home/thein/projects'
IGNORE_DIRS = [
    'node_modules', '.git', '.vscode', 'out', 'dist', 'build', 'coverage', '.cache'
]
IMPORTANT_FILES = [
    'package.json', 'README.md', '.augment-guidelines', 'tsconfig.json',
    'docker-compose.yml', 'Dockerfile', 'devcontainer.json'
]
MAX_BATCH_SIZE = 10  # Maximum number of entities/relations to create in a single batch

def read_file_safe(file_path: str, max_size: int = 10000) -> str:
    """Read file content safely with size limit."""
    try:
        file_size = os.path.getsize(file_path)
        if file_size > max_size:
            return f"File too large ({file_size} bytes)"
        
        with open(file_path, 'r', encoding='utf-8', errors='replace') as f:
            return f.read()
    except Exception as e:
        return f"Error reading file: {str(e)}"

def create_entities_batch(entities: List[Dict[str, Any]]) -> None:
    """Create entities in batches."""
    if not entities:
        return
    
    for i in range(0, len(entities), MAX_BATCH_SIZE):
        batch = entities[i:i+MAX_BATCH_SIZE]
        try:
            # Call the memory function to create entities
            print(f"Creating {len(batch)} entities (batch {i // MAX_BATCH_SIZE + 1})")
            print(json.dumps(batch, indent=2))
            # In a real implementation, this would be:
            # create_entities_memory({"entities": batch})
        except Exception as e:
            print(f"Error creating entities: {str(e)}")

def create_relations_batch(relations: List[Dict[str, Any]]) -> None:
    """Create relations in batches."""
    if not relations:
        return
    
    for i in range(0, len(relations), MAX_BATCH_SIZE):
        batch = relations[i:i+MAX_BATCH_SIZE]
        try:
            # Call the memory function to create relations
            print(f"Creating {len(batch)} relations (batch {i // MAX_BATCH_SIZE + 1})")
            print(json.dumps(batch, indent=2))
            # In a real implementation, this would be:
            # create_relations_memory({"relations": batch})
        except Exception as e:
            print(f"Error creating relations: {str(e)}")

def process_project(project_path: str) -> None:
    """Process a project and add it to the knowledge graph."""
    project_name = os.path.basename(project_path)
    print(f"Processing project: {project_name}")
    
    entities = []
    relations = []
    
    # Create project entity
    project_entity = {
        "name": project_name,
        "entityType": "Project",
        "observations": [f"Project directory: {project_path}"]
    }
    
    # Add package.json info if available
    package_json_path = os.path.join(project_path, 'package.json')
    if os.path.exists(package_json_path):
        try:
            package_json = json.loads(read_file_safe(package_json_path))
            if 'name' in package_json:
                project_entity["observations"].append(f"Name: {package_json['name']}")
            if 'description' in package_json:
                project_entity["observations"].append(f"Description: {package_json['description']}")
            if 'version' in package_json:
                project_entity["observations"].append(f"Version: {package_json['version']}")
        except Exception as e:
            print(f"Error parsing package.json for {project_name}: {str(e)}")
    
    # Add README.md content if available
    readme_path = os.path.join(project_path, 'README.md')
    if os.path.exists(readme_path):
        readme_content = read_file_safe(readme_path)
        project_entity["observations"].append(f"README: {readme_content}")
    
    entities.append(project_entity)
    
    # Process important files
    for file_name in IMPORTANT_FILES:
        file_path = os.path.join(project_path, file_name)
        if os.path.exists(file_path) and os.path.isfile(file_path):
            file_content = read_file_safe(file_path)
            entities.append({
                "name": f"{project_name}:{file_name}",
                "entityType": "File",
                "observations": [f"Path: {file_path}", f"Content: {file_content}"]
            })
            
            relations.append({
                "from": project_name,
                "to": f"{project_name}:{file_name}",
                "relationType": "contains"
            })
    
    # Process .devcontainer directory if it exists
    devcontainer_dir = os.path.join(project_path, '.devcontainer')
    if os.path.exists(devcontainer_dir) and os.path.isdir(devcontainer_dir):
        entities.append({
            "name": f"{project_name}:devcontainer",
            "entityType": "Component",
            "observations": ["DevContainer configuration for the project"]
        })
        
        relations.append({
            "from": project_name,
            "to": f"{project_name}:devcontainer",
            "relationType": "has"
        })
        
        # Process important devcontainer files
        devcontainer_files = ['devcontainer.json', 'docker-compose.yml', 'Dockerfile']
        for file_name in devcontainer_files:
            file_path = os.path.join(devcontainer_dir, file_name)
            if os.path.exists(file_path) and os.path.isfile(file_path):
                file_content = read_file_safe(file_path)
                entities.append({
                    "name": f"{project_name}:devcontainer:{file_name}",
                    "entityType": "File",
                    "observations": [f"Path: {file_path}", f"Content: {file_content}"]
                })
                
                relations.append({
                    "from": f"{project_name}:devcontainer",
                    "to": f"{project_name}:devcontainer:{file_name}",
                    "relationType": "contains"
                })
    
    # Create entities and relations
    create_entities_batch(entities)
    create_relations_batch(relations)

def build_knowledge_graph() -> None:
    """Main function to build the knowledge graph."""
    print("Building knowledge graph of projects...")
    
    try:
        # Get list of projects
        projects = [
            os.path.join(PROJECTS_DIR, item) for item in os.listdir(PROJECTS_DIR)
            if os.path.isdir(os.path.join(PROJECTS_DIR, item)) and item not in IGNORE_DIRS
        ]
        
        print(f"Found {len(projects)} projects: {', '.join(os.path.basename(p) for p in projects)}")
        
        # Process each project
        for project in projects:
            process_project(project)
        
        print("Knowledge graph built successfully!")
    except Exception as e:
        print(f"Error building knowledge graph: {str(e)}")

if __name__ == "__main__":
    build_knowledge_graph()
